test/metabase/test/redefs.clj

```diff
@@ -21,18 +21,38 @@
   (test-runner.parallel/assert-test-is-not-parallel "with-temp")
   ;; catch any Exceptions thrown when creating the object and rethrow them with some extra context to make them a
   ;; little easier to debug.
-  (let [temp-object (try
-                      (db/insert! model (merge (tt/with-temp-defaults model)
-                                               attributes))
-                      (catch Throwable e
-                        (throw (ex-info (str "with-temp error: " (ex-message e))
-                                        {:model (name model), :attributes attributes}
-                                        e))))
-        primary-key (models/primary-key model)]
+  (let [attributes-with-defaults (merge (tt/with-temp-defaults model)
+                                        attributes)]
     (try
-      (f temp-object)
-      (finally
-        (db/delete! model primary-key (primary-key temp-object))))))
+      (let [temp-object        (let [object (db/insert! model attributes-with-defaults)]
+                                 (assert (record? object)
+                                         (str (format "db/insert! for %s did not return a valid row. Got: %s."
+                                                      (name model)
+                                                      (pr-str object))
+                                              \newline
+                                              "(Tip: db/insert! doesn't seem to work with H2 with tables with compound PKs.)"))
+                                 object)
+            primary-key-column (models/primary-key model)
+            primary-key-value  (get temp-object primary-key-column)]
+        (assert primary-key-value (format "No value for primary key %s for row %s"
+                                          (pr-str primary-key-column)
+                                          (pr-str temp-object)))
+        (try
+          (f temp-object)
+          (finally
+            (db/delete! model primary-key-column primary-key-value))))
+      (catch Throwable e
+        ;; only wrap `e` if it's not another `with-temp` error. This way we don't get "false positives" if something
+        ;; fails at the beginning of a big `with-temp*` form for all the other models
+        (throw (if (::with-temp-error? (ex-data e))
+                 e
+                 (ex-info (format "with-temp error for %s: %s" (name model) (ex-message e))
+                          {:model                    (name model)
+                           :attributes               attributes
+                           :attributes-with-defaults attributes-with-defaults
+                           :primary-key-column       (models/primary-key model)
+                           ::with-temp-error?        true}
+                          e)))))))
```

Comment:
I added more validation and error handling here. I was banging my head against a wall trying to figure out why `with-temp` didn't work for `QueryAction` and `CardEmitter`, but only for H2. Turns out it was the aforementioned insert-returning-keys issues with `clojure.java.jdbc` and H2. I added this code here to make sure we catch Tables that do this so we can write stuff in a different way. It also includes more/better error messages

Reply:
Yeah I think doing that would make sense. Altho I don't think this stuff gets touched a lot I think I'm the only one who has ever touched this file and this is maybe the 3rd time